src/Data/Tree/AVL/Insertion.hs

```diff
@@ -21,21 +24,28 @@ import Data.Tree.AVL.Zipper
 
 --import qualified Debug.Trace as Debug
 
--- | Endpoint that allows to merge proofs for some sequental operations.
-insert' :: Stores h k v m => k -> v -> Map h k v -> m (Set h, Map h k v)
-insert' k v tree = do
-    ((), res, trails) <- runZipped' (insertZ k v) UpdateMode tree
+-- | Inserts given value for given key into the 'Map', generates proof prefab.
+--
+--   It is idempotent in terms of 'Map' content, however, without 'Eq' @k@
+--   constraint 'Map's will be different.
```

Comment:
Could you explain to me what it means?

Reply:
This is true (nice test, btw):
```haskell
quickcheck $ \k v tree ->
    Map.toList (Map.insert k v tree)
      ==
    Map.toList (Map.insert k v (Map.insert k v tree))
```

But this will surely fail:
```haskell
quickcheck $ \k v tree ->
    Map.insert k v tree
      ==
    Map.insert k v (Map.insert k v tree)
```

Thats because w/o `Eq v` we have no way to see if we're inserting the same thing into the tree.
_Unless_ I implement `(==)` for `AVL.Map` as hash-comparison.

Reply:
Mm, I'm still missing something.
1. Assuming `k :: K` and `v :: V`, since `toList` returns `[(K, V)]`, I still need `Eq v` for that comparison in the first case.
2. In second case, don't you just compare keys inside `insert`, and if the key is already present you overwrite the value (without comparing values)? (I may understand AVL interface bad and assume strange things).